src/runtime2/loading-indicator/service.js

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  CanPlay,
+  LoadPendingState,
+  LoadedState,
+  LoadingStrategy,
+  LoadingStrategyProp,
+  observeInput,
+  useDisposableMemo,
+  useMountComponent,
+  useRef,
+  useUnmountComponent,
+  withMetaData,
+} from '../../context';
+import {createLoaderElement} from '../../loader';
+import {devAssert} from '../../log';
+import {htmlFor} from '../../static-template';
+import {isInternalElement, isLoadingAllowed} from '../../layout';
+
+/**
+ * @param {!Node} root
+ * @return {!Function}
+ */
+export function LoadingIndicatorService(root) {
+  // console.log('LoadingIndicatorService: init');
+
+  const mountComponent = useMountComponent();
+  const unmountComponent = useUnmountComponent();
+
+  const io = new IntersectionObserver((records) => {
+    records.forEach((intersection) => {
+      const {target} = intersection;
+      mountComponent(LoadingIndicator, intersection, target);
+    });
+  });
+
+  const LoadPendingTracker = withMetaData(
+    [CanPlay, LoadingStrategyProp, LoadPendingState, LoadedState],
+    (node, _, canPlay, loadingStrategy, loadPending, loaded) => {
+      // console.log('LoadingIndicatorService: LoadPendingTracker: ',
+      //   node.nodeName + '#' + node.id,
+      //   canPlay, loadingStrategy, loadPending, loaded);
+      const indicatorsEnabled =
+        loadingStrategy == LoadingStrategy.AUTO ||
+        loadingStrategy == LoadingStrategy.LAZY ||
+        loadingStrategy == LoadingStrategy.EAGER;
+      if (indicatorsEnabled && canPlay && loadPending && !loaded) {
+        io.observe(node);
```

Comment:
Yeah. This might be the right path forward. I'm trying to look for more cases where input is needed. But TBH all of these concepts are in play now and depends on what we will see more often. For instance:

1. We may have only a handful of inputs and it'd make sense to remove that concept.
2. We might see a lot of optional props, in which case we'd be better off supporting a version  of `useContext()`. Which would make components re-run more often.

Also, I think we can also not stress too much about about "no input" case and just use the component as one big effect. It's fully reenterable and thus even executing it twice would not be a problem.